README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/clusteredinterference)](https://cran.r-project.org/package=clusteredinterference)
[![Travis-CI Build
Status](https://travis-ci.org/BarkleyBG/clusteredinterference.svg?branch=master)](https://travis-ci.org/BarkleyBG/clusteredinterference)
[![Coverage
Status](https://codecov.io/github/BarkleyBG/clusteredinterference/coverage.svg?branch=master)](https://codecov.io/github/BarkleyBG/clusteredinterference)

# About the ‘clusteredinterference’ package

This package implements the estimators proposed in [Barkley et al.
(2017), *Causal Inference from Observational Studies with Clustered
Interference*](https://arxiv.org/abs/1711.04834) for estimating the
causal effects of different treatment policies in the presence of
partial or clustered interference.

## What is clustered interference?

### What is interference?

In causal inference, when one individual’s treatment may affect another
individual’s outcome, it’s often called **interference**. In most
applications, it is assumed that there is no interference whatsoever. In
some applications this must be relaxed - e.g., as in infectious disease
research.

### Clustered interference is partial interference

A relaxation of the assumption of “no interference” is to assume that
individuals may be partitioned into distinct clusters of individuals
(e.g., households, or classrooms, etc.) such that there may be
interference within the clusters, but not between the clusters.
Historically, this assumption has been referred to as **partial
interference** after Sobel (2006).

[Barkley et al. (2017)](https://arxiv.org/abs/1711.04834) introduces the
terminology **clustered interference** to refer to this same assumption.
This phrase may be sufficiently descriptive of the underlying
assumption, and perhaps clarifies the presumed restriction of
interference to clusters.

# About the method

[Barkley et al. (2017)](https://arxiv.org/abs/1711.04834) proposes new
causal estimands for defining treatment effects in the context of
observational studies when there may be interference or spillover
effects between units in the same cluster. The manuscript also
introduces IPTW estimators for thos estimands, which are implemented in
‘clusteredinterference’.

## The manuscript

A version of this manuscript is available [on arXiv
at 1711.04834](https://arxiv.org/abs/1711.04834):

Barkley, B. G., Hudgens, M. G., Clemens, J. D., Ali, M., and Emch, M. E.
(2017). Causal inference from observational studies with clustered
interference. *arXiv preprint arXiv:1711.04834*. URL
<https://arxiv.org/abs/1711.04834>.

# Using the ‘clusteredinterference’ package

## Install the package

This package is now on CRAN\!

``` r
install.package("clusteredinterference")
```

Or, visit the [GitHub
repo:](https://github.com/BarkleyBG/clusteredinterference)

``` r
# devtools::install_github("BarkleyBG/clusteredinterference")
```

## Load this package

``` r
library(clusteredinterference)
```

## A quick data example

``` r
data("toy_data")
head(toy_data)
#>   Outcome Treatment Cluster_ID      Age Distance
#> 1       1         0          1 37.62826 6.485258
#> 2       0         1          1 36.61508 6.928957
#> 3       1         1          1 31.74776 6.659470
#> 4       1         0          2 34.79259 8.138802
#> 5       0         0          2 48.05607 7.736209
#> 6       0         0          2 42.21215 8.023865
```

## Example

Estimation is carried out with one function:

``` r
set.seed(1113)
causal_fx <- policyFX(
  data = toy_data,
  formula = Outcome | Treatment ~ Age + Distance + (1 | Cluster_ID) | Cluster_ID,
  alphas = c(.15, .25), 
  k_samps = 1
)
```

The estimates of causal estimands are printed in a tidy dataframe:

``` r
causal_fx
#> ------------- causal estimates --------------
#>       estimand estimate     se     LCI    UCI
#>       mu(0.15)   0.6985 0.0893  0.5234 0.8736
#>       mu(0.25)   0.6664 0.0702  0.5287 0.8041
#>      mu0(0.15)   0.7157 0.0917  0.5360 0.8954
#>      mu0(0.25)   0.6869 0.0775  0.5350 0.8388
#>      mu1(0.15)   0.1619 0.0429  0.0779 0.2460
#>      mu1(0.25)   0.2440 0.0536  0.1389 0.3491
#>  OE(0.25,0.15)  -0.0321 0.0275 -0.0861 0.0219
#>  OE(0.15,0.25)   0.0321 0.0275 -0.0219 0.0861
#>           ... and 4 more rows ... 
#> ---------------------------------------------
```

Use `summary()` for a little more information:

``` r
summary(causal_fx)
#> ------------- causal estimates --------------
#>       estimand estimate     se     LCI    UCI
#>       mu(0.15)   0.6985 0.0893  0.5234 0.8736
#>       mu(0.25)   0.6664 0.0702  0.5287 0.8041
#>      mu0(0.15)   0.7157 0.0917  0.5360 0.8954
#>      mu0(0.25)   0.6869 0.0775  0.5350 0.8388
#>      mu1(0.15)   0.1619 0.0429  0.0779 0.2460
#>      mu1(0.25)   0.2440 0.0536  0.1389 0.3491
#>  OE(0.25,0.15)  -0.0321 0.0275 -0.0861 0.0219
#>  OE(0.15,0.25)   0.0321 0.0275 -0.0219 0.0861
#> 
#>           ... and 4 more rows ... 
#> 
#> -------------- treatment model -------------
#> Generalized linear mixed model fit by maximum likelihood (Adaptive
#>   Gauss-Hermite Quadrature, nAGQ = 2) [glmerMod]
#>  Family: binomial  ( logit )
#> Formula: Treatment ~ Age + Distance + (1 | Cluster_ID)
#>    Data: data
#>      AIC      BIC   logLik deviance df.resid 
#> 137.0345 147.3743 -64.5172 129.0345       94 
#> Random effects:
#>  Groups     Name        Std.Dev.
#>  Cluster_ID (Intercept) 1.18    
#> Number of obs: 98, groups:  Cluster_ID, 30
#> Fixed Effects:
#> (Intercept)          Age     Distance  
#>    -1.44609     -0.00851      0.26097  
#> 
#> ------------- propensity scores -------------
#>      1      2      3      4      5      6      7      8      9     10 
#>  0.105  0.162  0.086  0.102  0.167  0.045  0.244 0.0934 0.0765  0.197 
#>     11     12     13     14     15     16     17     18     19     20 
#> 0.0653  0.281  0.104  0.365 0.0867  0.198  0.207  0.106 0.0847  0.134 
#>     21     22     23     24     25     26     27     28     29     30 
#>  0.103  0.111  0.105  0.302 0.0434 0.0943 0.0443 0.0512   0.13  0.263 
#> ---------------------------------------------
```

Note that `Treatment ~ Age + Distance + (1 | Cluster_ID)` in the the
middle of the `formula` argument is sent to `lme4::glmer()` to specify
the form of the (logit-link binomial) treatment model.

The `policyFX()` output list includes an element, `formula`, for the
`Formula` object:

``` r
causal_fx$formula
#> Outcome | Treatment ~ Age + Distance + (1 | Cluster_ID) | Cluster_ID
```

The output list also includes an element, `model`, which is the fitted
`glmerMod` S4 model object. Here we can see that the middle of `formula`
was passed into the `glmer()` logit-link binomial mixed model:

``` r
causal_fx$model@call
#> lme4::glmer(formula = Treatment ~ Age + Distance + (1 | Cluster_ID), 
#>     data = data, family = stats::binomial, nAGQ = nAGQ)
```

The fitted model estimates three fixed effects (intercept, a term for
`Age` and a term for `Distance`) and one random effect (for
`Cluster_ID`):

``` r
lme4::getME(causal_fx$model, c("beta", "theta"))
#> $beta
#> [1] -1.44608710 -0.00850977  0.26096895
#> 
#> $theta
#> Cluster_ID.(Intercept) 
#>               1.180325
```

## Vignette

The vignette provides more information on the formal arguments:

``` r
vignette("estimate-policyFX")
```

# News and version history

A changelog is found in the `NEWS.md` file. Version history is also
tracked by the [release
tags](https://github.com/BarkleyBG/clusteredinterference/releases) for
this GitHub repo.

# References and acknowledgments

  - The manuscript introducing the methods in ‘clusteredinterference’
    is:
      - Barkley, B. G., Hudgens, M. G., Clemens, J. D., Ali, M., and
        Emch, M. E. (2017). Causal inference from observational studies
        with clustered interference. *arXiv preprint arXiv:1711.04834*.
        URL <https://arxiv.org/abs/1711.04834>.
  - The terminology of **partial interference** is attributed to Sobel
    (2006):
      - Sobel, M. E. (2006). What do randomized studies of housing
        mobility demonstrate? Causal inference in the face of
        interference. *Journal of the American Statistical Association*,
        101(476), 1398-1407.
        [doi: 10.1198/016214506000000636](http://dx.doi.org/10.1198/016214506000000636)
  - Please see the
    [`inferference`](https://cran.r-project.org/package=inferference)
    package for related estimators from the following articles:
      - Perez‐Heydrich, C., Hudgens, M. G., Halloran, M. E., Clemens, J.
        D., Ali, M., & Emch, M. E. (2014). Assessing effects of cholera
        vaccination in the presence of interference. *Biometrics*,
        70(3), 731-741. [doi:
        10.1111/biom.12184](doi.wiley.com/10.1111/biom.12184)
      - Tchetgen, E. J. T., & VanderWeele, T. J. (2012). On causal
        inference in the presence of interference. *Statistical Methods
        in Medical Research*, 21(1), 55-75.
        [doi: 10.1177/0962280210386779](https://doi.org/10.1177/0962280210386779)  
  - An earlier version of the methods implemented in
    ‘clusteredinterference’ was implemented using the
    [`geex`](https://github.com/bsaul/geex) package for estimating
    equations.
  - Thanks to [Bradley Saul](https://github.com/bsaul) for
    [`inferference`](https://cran.r-project.org/package=inferference),
    [`geex`](https://github.com/bsaul/geex), and for comments and
    suggestions that were helpful in the creation of
    ‘clusteredinterference’.


NEWS.md
# clusteredinterference 1.0.0.9000

# clusteredinterference 1.0.0

- Initial CRAN release of 'clusteredinterference' package!
    - This package implements the estimators proposed in [Barkley et al. (2017), _Causal Inference from Observational Studies with Clustered Interference_](https://arxiv.org/abs/1711.04834) for estimating the causal effects of different treatment policies in the presence of partial or clustered interference.
- See the estimation function `policyFX()` for more information.



R/Estimator.R

#' Estimate Causal Effects of Population Treatment Policies Assuming Clustered
#' Interference
#'
#' This function implements the estimators from Barkley et al. (2017) for
#' estimating causal effects \emph{("FX")} of treatment policies from an
#' observational study when \strong{clustered interference} is assumed.
#' Clustered interference is also often known as "partial" interference. For the
#' manuscript introducing the methods in \pkg{clusteredinterference}, see: URL
#' \url{https://arxiv.org/abs/1711.04834}
#'
#'
#' @param data A \code{data.frame} (not a \code{tibble}). Columns of
#'   \code{factor} types are not recommended and will sometimes throw
#'   (defensive) errors.
#' @param formula The \code{formula} defines the different components of the
#'   method. The components are specified by \code{outcome | treatment ~
#'   f(covariates) + (1|cluster_id) | cluster_id}. The middle component is
#'   passed to \code{\link[lme4]{glmer}}, so \code{treatment ~ f(covariates) +
#'   (1|cluster_id)} specifies the model form for the propensity score (i.e.,
#'   treatment) model. See Details.
#' @param alphas A numeric vector for the probabilities corresponding to the
#'   policies of interest. Each entry must be between 0 and 1.
#' @param k_samps The maximum number of vectors to evaluate to estimate the
#'   counterfactual probabilities (i.e., \eqn{\omega(A,N,\alpha)}). Setting to 0
#'   avoids approximation at the cost of increased computation time. Recommended
#'   to set <= 5.
#' @param ... The dots argument. The user may supply their own
#'   \code{target_grid} through the dots argument. The \code{target_grid} can be
#'   made through exported function \code{\link{makeTargetGrid}}
#' @param nAGQ This is the number of Adaptive Gaussian Quadrature points used in
#'   the \code{\link[lme4]{glmer}} model fitting computation. Defaults to 2. It
#'   is recommended to use more than 1.
#' @param root_options These are passed to \code{\link[rootSolve]{multiroot}}
#'   function.
#' @param return_matrices A Boolean on whether to return the "bread" and "meat"
#'   matrices in the sandwich variance. Defaults to \code{FALSE}.
#' @param verbose A Boolean on whether to print output to \code{stderr}.
#'   Defaults to FALSE.
#'
#' @details These estimators are based on inverse probability-weighting by the
#'   propensity score for treatment (IPW) to estimate causal effects of
#'   counterfactual policies of interest (i.e., \emph{"policy effects"}) when
#'   clustered interference is assumed. The policies of interest correspond to
#'   counterfactual scenarios in which treatment may be correlated within
#'   clusters.
#'
#'   This method estimates causal contrasts of these policies by estimating the
#'   counterfactual treatment probabilities; taking the correlation structures
#'   into account requires heavy computational resources, so the user should be
#'   patient.
#'
#'   The modeling formula for the propensity score (i.e., treatment) model is
#'   specified via the \code{formula} formal argument. An example of a model
#'   logit-linear fixed effects would be \code{Y | A ~ X1 + X2 + (1 |
#'   cluster_ID) | cluster_ID}. A similar model that also includes an
#'   interaction term is \code{Y | A ~ X1 + X2 + X1:X2 + (1 | cluster_ID) |
#'   cluster_ID}.
#'
#' @author Brian G. Barkley, \email{BarkleyBG@@unc.edu}.
#'
#'   Some of the plumbing functions for estimating the sandwich variance matrix
#'   were adapted from Bradley Saul's
#'   \href{https://cran.r-project.org/package=geex}{\pkg{geex}} package.
#'
#'   Some of the plumbing functions for the logistic mixed model likelihood were
#'   adapted from Bradley Saul's
#'   \href{https://cran.r-project.org/package=inferference}{\pkg{inferference}}
#'   package.
#'
#' @return A \code{list} object including: \enumerate{ \item \code{estimates}: A
#'   tidy \code{data.frame} with columns \code{estimand}, \code{estimate},
#'   \code{var}, \code{se}, \code{LCI} and \code{UCI} for 95\% CI's, and more
#'   information. \item \code{parameters}: An untidy \code{list} of the point
#'   estimates of all (target and nuisance) parameters. \item
#'   \code{variance_matrices}: When \code{return_matrices} is \code{TRUE} this
#'   is a \code{list} object for the "bread" and "meat" matrices in the sandwich
#'   variance calculations for each estimand. Otherwise, it is a \code{list}
#'   object with length 0. \item \code{propensity_scores}: The estimated
#'   propensity scores for each cluster. \item \code{model}: The treatment model
#'   object. \item \code{formula}: The full formula argument provided, after
#'   coercion to a \code{\link[Formula]{Formula}} object}
#'
#' @seealso Please see the main package vignette at
#'   \code{vignette("estimate-policyFX")}. It describes the necessary arguments,
#'   as well as some extra functionality.
#'
#' @references Barkley, B. G., Hudgens, M. G., Clemens, J. D., Ali, M., and
#'   Emch, M. E. (2017). Causal Inference from Observational Studies with
#'   Clustered Interference. \emph{arXiv preprint arXiv:1711.04834}. (URL:
#'   \url{https://arxiv.org/abs/1711.04834}.)
#'
#'   Bradley C Saul and Michael G Hudgens (2017). A Recipe for
#'   \code{inferference}: Start with Causal Inference. Add Interference. Mix
#'   Well with R. \emph{Journal of Statistical Software} \strong{82}(2), pp.
#'   1-21. doi: <10.18637/jss.v082.i02> (URL:
#'   \url{http://doi.org/10.18637/jss.v082.i02}).
#'   \url{https://cran.r-project.org/package=inferference}.
#'   \url{https://github.com/bsaul/inferference}.
#'
#'   Bradley Saul (2017). \code{geex}: An API for M-Estimation.
#'   \url{https://cran.r-project.org/package=geex}.
#'   \url{https://github.com/bsaul/geex}.
#'
#'
#' @author Brian G. Barkley

#' @examples
#' \dontrun{
#' data("toy_data", "clusteredinterference")
#' causal_fx <- policyFX(
#'   data = toy_data,
#'   formula = Outcome | Treatment ~ Age + Distance + (1 | Cluster_ID) | Cluster_ID,
#'   alphas = c(.3, .5),
#'   k_samps = 1,
#'   verbose = FALSE
#' )}
#' @export
policyFX <- function(
  data,
  formula,
  alphas,
  k_samps=NULL,
  ...,
  verbose = FALSE,
  root_options = NULL,
  nAGQ=2,
  return_matrices = FALSE
){

  if (is.null(k_samps)) {
    stop("Please specify a numeric for `k_samps` argument. 0 is for full sample, but <=5 is recommended")
  }
  max_k_evals <- k_samps

  dots <- list(...)
  if ("target_grid" %in% names(dots)) {
    target_grid <- dots$target_grid
    alphas_in_grid <- sort(unique(target_grid$alpha1))
    if (missing("alphas")){
      alphas <- alphas_in_grid
    }
    if (!all(alphas == alphas_in_grid)){
      warning("target_grid has different alphas than those specified; the alphas from the target_grid argument will be used")
      alphas <- alphas_in_grid
    }
  } else {
    target_grid <- makeTargetGrid(alphas = alphas)
  }

  target_grid$k <- max_k_evals

  formule <- Formula::Formula(formula)

  arguments <- argTests(
    data = data,
    formule = formule,
    alphas = alphas,
    max_k_evals = max_k_evals
  )

  data <- arguments$data ##sorted
  alphas <- arguments$alphas ##sorted
  modeling_formula <- arguments$modeling_formula
  grouping_var_name<- arguments$grouping_var_name

  ## verbose
  if (verbose){print("I. Estimate GLMER model")}
  ## verbose

  glmer_fit <- lme4::glmer(
    data = data,
    formula = modeling_formula,
    family = stats::binomial,
    nAGQ = nAGQ
  )

  model_parms <- unlist(lme4::getME(glmer_fit, c("beta","theta")))


  split_data_list <- splitData(
    data = data,
    formule = formule,
    glmer_fit = glmer_fit,
    grouping_var_name = grouping_var_name
  )

  ## verbose
  if (verbose){print("IB. get MCFP vecs")}
  ## verbose

  vec_dfm_no_alphas <- sampleVecs4Trt(
    split_data_list = split_data_list,
    max_k_evals = max_k_evals
  )


  index_references <- makeIndexReferences(
    model_parms = model_parms,vec_dfm_no_alphas = vec_dfm_no_alphas)

  typical_args <- list(
    split_data_list = split_data_list,
    model_parms = model_parms,
    verbose = verbose
  )
  CFBI_args <- append(typical_args, list(
    alphas = alphas,
    glmer_fit = glmer_fit,
    root_options = root_options,
    vec_dfm_no_alphas = vec_dfm_no_alphas
  ))
  CFBI_MCFP_list <- do.call(what=estimateNuisance,args=CFBI_args)

  CFBI_estimates <-  CFBI_MCFP_list$CFBI_estimates
  MCFP_tidy_estimates <- CFBI_MCFP_list$MCFP_tidy_estimates


  target_args <- append(typical_args, list(
    CFBI_estimates = CFBI_estimates,
    MCFP_tidy_estimates = MCFP_tidy_estimates,
    alphas = alphas,
    target_grid = target_grid
  ))
  ## to output
  targets_and_cps <- do.call(what=estimateTargets,args=target_args)


  target_estimates <- targets_and_cps$target_grid

  var_args <- target_args
  var_args$target_grid <- NULL
  var_args$target_grid_estimates <- target_estimates


  ## Diagnostics
  MCFP_tidier_estimates <- tidyOutMCFP(MCFP_tidy_estimates)

  ## to output
  parameters <-  list(
    model_parms = model_parms,
    CFBI_estimates = CFBI_estimates,
    MCFP_estimates = MCFP_tidier_estimates,
    target_estimates = target_estimates
  )
  if (verbose){print('Beginning variance calculations. This may take some time.')}

  var_args$index_references <- index_references
  var_args$return_matrices <- return_matrices



  var_ests_list <- do.call(estimateVariance, args = var_args)

  ## to ouptut
  tidy_estimates <- tidyEstimates(var_ests_list$tidy_grid)
  ## to output
  if (return_matrices== TRUE){
    variance_matrices <-  var_ests_list$var_matrices
  } else{
    variance_matrices <- list()
  }



  out_list <- list(
    estimates = tidy_estimates,
    parameters = parameters,
    variance_matrices = variance_matrices,
    propensity_scores = targets_and_cps$prop_scores,
    model = glmer_fit,
    formula = formule
  )
  class(out_list) <- "policyFX"
  out_list

}
